File: src/typescript/shared/keystore.ts
import * as argon2 from 'argon2';
import * as nacl from 'tweetnacl';
import { readFileSync, writeFileSync } from 'fs';
import { resolve } from 'path';
import { Keypair } from '@solana/web3.js';

const ARGON2_OPTIONS = {
  type: argon2.argon2id,
  timeCost: 4,
  memoryCost: 65536,
  parallelism: 2,
  hashLength: 32,
  raw: true, // Get raw bytes instead of formatted string
};

export interface KeystoreData {
  salt: string;
  nonce: string;
  encryptedSecretKey: string;
  kdfParams: {
    timeCost: number;
    memoryCost: number;
    parallelism: number;
  };
}

export const createKeystore = async (
  secretKey: Uint8Array,
  passphrase: string,
  keystorePath: string
): Promise<void> => {
  const salt = Buffer.from(nacl.randomBytes(16));
  const derivedKey = await argon2.hash(passphrase, { ...ARGON2_OPTIONS, salt });
  // derivedKey is already a Buffer, use it directly
  const key = Buffer.from(derivedKey);
  const nonce = nacl.randomBytes(nacl.secretbox.nonceLength);
  const encrypted = nacl.secretbox(secretKey, nonce, key);

  const keystoreData: KeystoreData = {
    salt: Buffer.from(salt).toString('hex'),
    nonce: Buffer.from(nonce).toString('hex'),
    encryptedSecretKey: Buffer.from(encrypted).toString('hex'),
    kdfParams: {
      timeCost: ARGON2_OPTIONS.timeCost,
      memoryCost: ARGON2_OPTIONS.memoryCost,
      parallelism: ARGON2_OPTIONS.parallelism,
    },
  };

  writeFileSync(keystorePath, JSON.stringify(keystoreData, null, 2), { mode: 0o600 });
};

export const loadKeypairFromKeystore = async (
  keystorePath: string,
  passphrase: string
): Promise<Keypair> => {
  const keystoreData: KeystoreData = JSON.parse(readFileSync(keystorePath, 'utf-8'));
  const salt = Buffer.from(keystoreData.salt, 'hex');
  const derivedKey = await argon2.hash(passphrase, { ...ARGON2_OPTIONS, salt });
  const key = Buffer.from(derivedKey); // Already a Buffer
  const nonce = Buffer.from(keystoreData.nonce, 'hex');
  const encrypted = Buffer.from(keystoreData.encryptedSecretKey, 'hex');

  const secretKey = nacl.secretbox.open(encrypted, nonce, key);
  if (!secretKey) throw new Error('Invalid passphrase or corrupted keystore');

  // FIX: Create a hard copy of the secret key bytes before passing to Keypair
  // This breaks the memory reference so fill(0) won't destroy the Keypair's secret
  const secretKeyCopy = new Uint8Array(secretKey);
  const keypair = Keypair.fromSecretKey(secretKeyCopy);
  
  // Now it is safe to clear the original decrypted buffer
  secretKey.fill(0);
  return keypair;
};

File: src/python/shared/keystore.py
import json
import nacl.secret
import nacl.utils
import argon2
from pathlib import Path
from typing import Dict, Any

ARGON2_OPTIONS = {
    "time_cost": 4,
    "memory_cost": 65536,
    "parallelism": 2,
    "hash_len": 32,
    "type": argon2.Type.ID,
}


class Keystore:
    def __init__(self, keystore_path: str):
        self.keystore_path = Path(keystore_path)
        if not self.keystore_path.exists():
            raise FileNotFoundError(f"Keystore not found at {keystore_path}")

    def load_keypair(self, passphrase: str) -> bytes:
        with open(self.keystore_path, "r") as f:
            data: Dict[str, Any] = json.load(f)

        salt = bytes.fromhex(data["salt"])
        nonce = bytes.fromhex(data["nonce"])
        encrypted_secret = bytes.fromhex(data["encryptedSecretKey"])

        derived_key = argon2.low_level.hash_secret_raw(
            passphrase.encode("utf-8"), salt, **ARGON2_OPTIONS
        )

        box = nacl.secret.SecretBox(derived_key)
        return box.decrypt(encrypted_secret, nonce)

    @staticmethod
    def create_keystore(secret_key: bytes, passphrase: str, keystore_path: str) -> None:
        salt = nacl.utils.random(16)
        derived_key = argon2.low_level.hash_secret_raw(
            passphrase.encode("utf-8"), salt, **ARGON2_OPTIONS
        )
        nonce = nacl.utils.random(nacl.secret.SecretBox.NONCE_SIZE)
        box = nacl.secret.SecretBox(derived_key)
        encrypted = box.encrypt(secret_key, nonce)

        keystore_data = {
            "salt": salt.hex(),
            "nonce": nonce.hex(),
            "encryptedSecretKey": encrypted.ciphertext.hex(),
            "kdfParams": {
                "timeCost": ARGON2_OPTIONS["time_cost"],
                "memoryCost": ARGON2_OPTIONS["memory_cost"],
                "parallelism": ARGON2_OPTIONS["parallelism"],
            },
        }

        with open(keystore_path, "w") as f:
            json.dump(keystore_data, f, indent=2)
        Path(keystore_path).chmod(0o600)
